sql: Close() called on nil StmtBuf

On master today

```
$ make stress PKG=./pkg/kv/kvserver/protectedts/ptcache/ TESTS=TestSettingChangedLeadsToFetch 
[10 seconds-ish pass]

panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0x311c009]

goroutine 512 [running]:
github.com/cockroachdb/cockroach/pkg/sql.(*StmtBuf).Close(0x0)
	/home/tobias/go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_io.go:387 +0x29
github.com/cockroachdb/cockroach/pkg/sql.(*InternalExecutor).execInternal.func3(0x4e05d20, 0xc0000ca280)
	/home/tobias/go/src/github.com/cockroachdb/cockroach/pkg/sql/internal.go:715 +0x3a
github.com/cockroachdb/cockroach/pkg/sql.(*InternalExecutor).initConnEx.func1(0xc002346000, 0x4e93780, 0xc002176a50, 0xc0005ae900, 0xc00216d020, 0xc00210cd80, 0xc001a4c7e0, 0xc001a10d20)
	/home/tobias/go/src/github.com/cockroachdb/cockroach/pkg/sql/internal.go:204 +0x153
created by github.com/cockroachdb/cockroach/pkg/sql.(*InternalExecutor).initConnEx
	/home/tobias/go/src/github.com/cockroachdb/cockroach/pkg/sql/internal.go:201 +0x356


ERROR: exit status 2

558 runs completed, 1 failures, over 9s
context canceled
FAIL
FAIL	github.com/cockroachdb/cockroach/pkg/kv/kvserver/protectedts/ptcache	9.230s
FAIL
make: *** [Makefile:1034: stress] Error 1
```

--- a/pkg/sql/internal.go
+++ b/pkg/sql/internal.go
@@ -125,8 +125,8 @@ func (ie *InternalExecutor) SetSessionData(sessionData *sessiondata.SessionData)
 }
 
 // initConnEx creates a connExecutor and runs it on a separate goroutine. It
-// returns a StmtBuf into which commands can be pushed and a WaitGroup that will
-// be signaled when connEx.run() returns.
+// takes in a StmtBuf into which commands can be pushed and a WaitGroup that
+// will be signaled when connEx.run() returns.
 //
 // If txn is not nil, the statement will be executed in the respective txn.
 //
@@ -139,9 +139,11 @@ func (ie *InternalExecutor) initConnEx(
 	txn *kv.Txn,
 	ch chan ieIteratorResult,
 	sd *sessiondata.SessionData,
+	stmtBuf *StmtBuf,
+	wg *sync.WaitGroup,
 	syncCallback func([]resWithPos),
 	errCallback func(error),
-) (*StmtBuf, *sync.WaitGroup) {
+) {
 	clientComm := &internalClientComm{
 		ch: ch,
 		// init lastDelivered below the position of the first result (0).
@@ -166,7 +168,6 @@ func (ie *InternalExecutor) initConnEx(
 	}
 	appStats := ie.s.sqlStats.getStatsForApplication(appStatsBucketName)
 
-	stmtBuf := NewStmtBuf()
 	var ex *connExecutor
 	if txn == nil {
 		ex = ie.s.newConnExecutor(
@@ -196,7 +197,6 @@ func (ie *InternalExecutor) initConnEx(
 	}
 	ex.executorType = executorTypeInternal
 
-	var wg sync.WaitGroup
 	wg.Add(1)
 	go func() {
 		if err := ex.run(ctx, ie.mon, mon.BoundAccount{} /*reserved*/, nil /* cancel */); err != nil {
@@ -211,7 +211,6 @@ func (ie *InternalExecutor) initConnEx(
 		ex.close(ctx, closeMode)
 		wg.Done()
 	}()
-	return stmtBuf, &wg
 }
 
 type ieIteratorResult struct {
@@ -631,8 +630,8 @@ func (ie *InternalExecutor) execInternal(
 	// exits before the span is finished.
 	ctx, sp := tracing.EnsureChildSpan(ctx, ie.s.cfg.AmbientCtx.Tracer, opName)
 
-	var stmtBuf *StmtBuf
-	var wg *sync.WaitGroup
+	stmtBuf := NewStmtBuf()
+	var wg sync.WaitGroup
 
 	defer func() {
 		// We wrap errors with the opName, but not if they're retriable - in that
@@ -645,20 +644,8 @@ func (ie *InternalExecutor) execInternal(
 			if !errIsRetriable(retErr) {
 				retErr = errors.Wrapf(retErr, "%s", opName)
 			}
-			if stmtBuf != nil {
-				// If stmtBuf is non-nil, then the connExecutor goroutine has
-				// been spawn up - we gotta wait for it to exit.
-				//
-				// Note that at the moment of writing when retErr is non-nil,
-				// the stmtBuf is necessarily nil (the only errors emitted after
-				// the connExecutor is initialized are the errors on pushing
-				// into the stmtBuf, and those could occur only if the stmtBuf
-				// is closed which would indicate problems with
-				// synchronization). In any case, we want to be safe and handle
-				// such a scenario accordingly.
-				stmtBuf.Close()
-				wg.Wait()
-			}
+			stmtBuf.Close()
+			wg.Wait()
 			sp.Finish()
 		} else {
 			// r must be non-nil here.
@@ -709,13 +696,15 @@ func (ie *InternalExecutor) execInternal(
 		}
 		ch <- ieIteratorResult{err: errors.AssertionFailedf("missing result for pos: %d and no previous error", resPos)}
 	}
+	// errCallback is called if an error is returned from the connExecutor's
+	// run() loop.
 	errCallback := func(err error) {
 		// The connExecutor exited its run() loop, so the stmtBuf must have been
 		// closed. Still, since Close() is idempotent, we'll call it here too.
 		stmtBuf.Close()
 		ch <- ieIteratorResult{err: err}
 	}
-	stmtBuf, wg = ie.initConnEx(ctx, txn, ch, sd, syncCallback, errCallback)
+	ie.initConnEx(ctx, txn, ch, sd, stmtBuf, &wg, syncCallback, errCallback)
 
 	typeHints := make(tree.PlaceholderTypes, len(datums))
 	for i, d := range datums {
@@ -768,7 +757,7 @@ func (ie *InternalExecutor) execInternal(
 		ch:         ch,
 		resultCols: resultColumns,
 		stmtBuf:    stmtBuf,
-		wg:         wg,
+		wg:         &wg,
 	}, nil
 }
 

